File: leetcode/Medium/3.LongestSubstringWithoutRepeatingCharacters.go
package main

import (
	"fmt"
	"math"
)

func lengthOfLongestSubstring(s string) int {
	size := len(s)
	st := make(map[byte]struct{})
	right := 0
	ans := 0
	for left := 0; left < size; left++ {
		for right < size {
			if _, ok := st[s[right]]; ok {
				break
			}
			st[s[right]] = struct{}{}
			right++
		}
		ans = int(math.Max(float64(ans), float64(len(st))))
		delete(st, s[left])
	}
	return ans
}

func main() {
	fmt.Println(lengthOfLongestSubstring("abcabcbb"))
}

// 21ms, 39.88%

File: leetcode/Medium/3.LongestSubstringWithoutRepeatingCharacters.cpp
#include <bits/stdc++.h>
using namespace std;
#define _GLIBCXX_DEBUG

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        const int len = s.size();
        set<char> st;
        unsigned int ans = 0; // submit: unsigned long
        int right = 0;
        for (int left = 0; left < len; left++) {
            while (right < len && st.find(s[right]) == st.end()) {
                st.insert(s[right]);
                right++;
            }
            ans = max(ans, st.size());
            st.erase(s[left]);
        }
        return (int)ans;
    }
};

int main(void) {
    Solution s;
    cout << s.lengthOfLongestSubstring("abcabcbb") << endl;
}

// 40ms, 45.4%
